python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,149 @@
+
+    n_properties = len(block1.properties) + len(block2.properties)
+
+    properties = Labels(
+        names=["properties"],
+        values=np.arange(n_properties, dtype=np.int32).reshape(-1, 1),
+    )
+
+    values = _dispatch.hstack([block2.values, block1.values])
+
+    result_block = TensorBlock(
+        values=values,
+        samples=block2.samples,
+        components=[],
+        properties=properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+
+        gradient2 = block2.gradient(parameter)
+
+        gradient_data = np.zeros((len(gradient1.samples), 3, n_properties))
+
+        # copy block1 gradient data over
+        gradient_data[:, :, : len(block1.properties)] = gradient1.data
+
+        # check where to put the block2 data
+        for i_grad, sample_grad in enumerate(gradient2.samples):
+            i_merged = gradient1.samples.position(sample_grad)
+            gradient_data[i_merged, :, len(block1.properties) :] = gradient2.data[
+                i_grad, :, :
+            ]
+
+        result_block.add_gradient(
+            parameter,
+            gradient_data,
+            gradient1.samples,
+            gradient1.components,
+        )
+
+    return result_block
+
+
+def _join_along_samples_block(block1: TensorBlock, block2: TensorBlock) -> TensorBlock:
+    for property1, property2 in zip(block1.properties, block2.properties):
+        if tuple(property1) != tuple(property2):
+            raise ValueError("`A` and `B` do not have the same properties.")
+
+    if len(block1.components) > 0:
+        raise ValueError("`A` should not have any components")
+
+    if len(block2.components) > 0:
+        raise ValueError("`B` should not have any components")
+
+    n_samples = len(block1.samples) + len(block2.samples)
+
+    samples = Labels(
+        names=["samples"], values=np.arange(n_samples, dtype=np.int32).reshape(-1, 1)
+    )
+
+    values = _dispatch.vstack([block2.values, block1.values])
+
+    result_block = TensorBlock(
+        values=values,
+        samples=samples,
+        components=[],
+        properties=block1.properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+
+        gradient2 = block2.gradient(parameter)
+
+        n_samples_gradient = len(gradient1.samples) + len(gradient2.samples)
+        gradient_data = np.zeros((n_samples_gradient, 3, len(gradient1.properties)))
+
+        gradient_data[: len(gradient1.samples), :, :] = gradient1.data
+        gradient_data[len(gradient1.samples) :, :, :] = gradient2.data
+
+        gradients_samples = Labels(
+            names=["gradients_samples"],
+            values=np.arange(n_samples_gradient, dtype=np.int32).reshape(-1, 1),
+        )
+
+        # TODO: calls to add_gradient are failing...
+        result_block.add_gradient(
+            parameter,
+            gradient_data,
+            gradients_samples,
+            gradient1.components,
+        )
```

Comment:
I am a bit puzzled here. This leads to

```
EquistoreError: invalid parameter: first variable in the gradients samples labels must be 'samples'
```

but I constructed the gradient sample labels in the same way as the other labels.

Reply:
the names in the gradient samples must be `["sample", "something", "else"]`, e.g. positions gradients use `["sample", "structure", "atom"]`, and cell gradients use `["sample"]`. 

Here you are passing `["gradients_samples"]` as the names. You need instead to update the gradient samples (some of them will have to change, because they refer to the indices of the samples in the initial blocks)

---

BTW, the error message is misleading, it should be `sample` not `samples`, and it could include what the user gave for more context.

Reply:
Ahh I see. So in principle we should update all sample and property names, right?

Reply:
when joining along samples; you should not have to update sample names, but the joined tensormap should have the same sample names already. You will have to update the gradient sample **values**, and ensure the values samples are distinct.

---

When joining along properties, you can either try to keep the metdata from the previous properties (if they have the same names & are distinct, take the union) or re-create new properties with new names. Maybe something like 

```
properties_1 = ["n"], [0, 2, 3]   |=> properties = ["n"], [0, 2, 3, 1, 4, 5]
properties_2 = ["n"], [1, 4, 5]   |

properties_1 = ["n"], [0, 2, 3]   |=> properties = ["tensor", "n"], [[0, 0], [0, 2], [0, 3], [1, 0], [1, 2]]
properties_2 = ["n"], [0, 2]      |

properties_1 = ["a"], [0, 2, 3]             |=> properties = ["tensor", "q"], [[0, 0], [0, 1], [0, 2], [1, 0], [1, 1]]
properties_2 = ["b", "c], [[0, 0], [1, 2]]  |
```

in the last case, we loose some information, which could be preserved if we add arbitrary string metadata to tensormaps (https://github.com/lab-cosmo/equistore/blob/fa1db6410da7a1eba995e98e6b30843905bdeae6/src/tensor/mod.rs#L30)

Reply:
I changed the behavior of setting the samples and properties. Let me know if you agree that the logic makes sense @Luthaf. If yes I will deal with the open todo and add tests.

Reply:
I'm not sure I can follow the logic without comments or running it on some examples to be honest =). I'll check this branch out locally

Reply:
Yes it is a bit hard so see. While I am writing the tests it seems that there are still some things I have to sort out.

Reply:
Okay now it should work and you can try some examples. I will do the tests tomorrow.